p2p: index TxOrphanage by wtxid, allow entries with same txid

glozow opened the pull request:
Part of #27463 in the "make orphan handling more robust" section.

Currently the main map in `TxOrphanage` is indexed by txid; we do not allow 2 transactions with the same txid into TxOrphanage. This means that if we receive a transaction and want to store it in orphanage, we'll fail to do so if a same-txid-different-witness version of the tx already exists in the orphanage. The existing orphanage entry can stay until it expires 20 minutes later, or until we find that it is invalid.

This means an attacker can try to block/delay us accepting an orphan transaction by sending a mutated version of the child ahead of time. See included test.

Prior to #28970, we don't rely on the orphanage for anything and it would be relatively difficult to guess what transaction will go to a node's orphanage. After the parent(s) are accepted, if anybody sends us the correct transaction, we'll end up accepting it. However, this is a bit more painful for 1p1c: it's easier for an attacker to tell when a tx is going to hit a node's orphanage, and we need to store the correct orphan + receive the parent before we'll consider the package. If we start out with a bad orphan, we can't evict it until we receive the parent + try the 1p1c, and then we'll need to download the real child, put it in orphanage, download the parent again, and then retry 1p1c.

glozow committed:
[functional test] attackers sending mutated orphans

glozow committed:
[p2p] allow entries with the same txid in TxOrphanage

Index by wtxid instead of txid to allow entries with the same txid but
different witnesses in orphanage. This prevents an attacker from
blocking a transaction from entering the orphanage by sending a mutated
version of it.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -2295,7 +2295,9 @@ bool PeerManagerImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconside
-    if (m_orphanage.HaveTx(gtxid)) return true;
+    // Orphanage is checked by wtxid. However, even if this is a txid, look up the same hash in
+    // case this is a non-segwit transaction in the orphanage.
+    if (m_orphanage.HaveTx(Wtxid::FromUint256(gtxid.GetHash()))) return true;
```
```Suggestion
    if (m_orphanage.HaveTx(Wtxid::FromUint256(hash))) return true;
```

glozow replied:
done

instagibbs reviewed (commented):
code changes are quite mechanical, thankfully

concept ACK

instagibbs commented:
should be able to add fuzz coverage easily for differing wtxids but same txids in `src/test/fuzz/txorphan.cpp`

stickies-v commented on src/net_processing.cpp:
```diff
@@ -2295,7 +2295,9 @@ bool PeerManagerImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconside
     const uint256& hash = gtxid.GetHash();
 
-    if (m_orphanage.HaveTx(gtxid)) return true;
+    // Orphanage is checked by wtxid. However, even if this is a txid, look up the same hash in
+    // case this is a non-segwit transaction in the orphanage.
+    if (m_orphanage.HaveTx(Wtxid::FromUint256(gtxid.GetHash()))) return true;
```
It's not immediately obvious to me why it is safe to just convert any `gtxid` into a `Wtxid` here, especially since there are callsites of `AlreadyHaveTx` where a ` GenTxId::Txid` is passed, such as e.g. [here](https://github.com/bitcoin/bitcoin/blob/d73245abc70346a0e8805d50a1f395706084294c/src/net_processing.cpp#L4632). If it is indeed safe, I think it could be a better approach to update `AlreadyHaveTx` to take a `Wtxid` and push the conversion further to the edge, so it can be reviewed on a case-by-case basis?

AngusP replied:
(Ignore me if I'm understanding, still picking things up 😅) -- coinbase transactions are [required to have all `0` Wtxids](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#commitment-structure) (`0x0000....0000`) which means for an `INV` message for a witness coinbase you have to use the non-witness txid to reference the transaction? - so `AlreadyHaveTx` in that case is explicitly *not* using a Wtxid? (Or are coinbases never `INV`'d?)

Also perhaps not great/confusing to treat Wtxid and Txids as always interchangeable (though the types allow it) because of the coinbase Wtxid caveat?

glozow replied:
> It's not immediately obvious to me why it is safe to just convert any gtxid into a Wtxid here

The argument I'm making in this PR is that in all the places we are checking by txid, we shouldn't be doing that, because we're missing same-txid-different-witness cases. But yeah totally fair to say it's difficult to review it together, so I'll split the first commit to make it more explicit where the change is happening and why that's ok.

>  I think it could be a better approach to update AlreadyHaveTx to take a Wtxid

fwiw I disagree, as we do want to check txids in the other data structures.

> coinbase transactions are [required to have all 0 Wtxids](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#commitment-structure) (0x0000....0000) which means for an INV message for a witness coinbase you have to use the non-witness txid to reference the transaction? - so AlreadyHaveTx in that case is explicitly not using a Wtxid? (Or are coinbases never INV'd?)

That link is referring to the wtxid of the coinbase transaction when calculating the witness commitment; that's not the case in any other situation. Also, yes, if we get a coinbase transaction in tx relay we'd drop that pretty much immediately and wouldn't be putting it in orphanage:
https://github.com/bitcoin/bitcoin/blob/9d1a286f20b8a602ffe72928bcd79be09fdbf9d0/src/validation.cpp#L735-L742

stickies-v replied:
> The argument I'm making in this PR is that in all the places we are checking by txid, we shouldn't be doing that, because we're missing same-txid-different-witness cases.

I agree, the rationale makes sense. My concern is about segwit txs (`txid != wtxid`) where we check by txid (for whatever reason: intentional because we don't have witness data, legacy and we never updated it, buggy, ...): just creating a wtxid from the txid hash as is done here seems like it has potential to start missing things we previously wouldn't - since `TxOrphanage` now _only_ indexes by `wtxid`?

I've made a branch where I've created 2 overloads of `AlreadyHaveTx`: one that takes `Wtxid` and one that takes `Txid`, to force all the conversions to the edges. I implemented the `Txid` to perform less checks, based on the documentation (didn't verify the code) that `m_recent_rejects_reconsiderable` and `m_recent_rejects` uses `wtxid`, whereas `m_recent_confirmed_transactions` uses both `txid` and `wtxid`.

There are 4 callsites of `AlreadyHaveTx`. For 2 of them, [1) here](https://github.com/glozow/bitcoin/compare/2024-04-orphan-use-wtxid...stickies-v:bitcoin:pr/30000#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R4255-R4256) and [2) here](https://github.com/glozow/bitcoin/compare/2024-04-orphan-use-wtxid...stickies-v:bitcoin:pr/30000#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R4561), the conversion seems trivial/sensible.

A third is more confusing to me. [3) Here in `ProcessMessage`](https://github.com/glozow/bitcoin/compare/2024-04-orphan-use-wtxid...stickies-v:bitcoin:pr/30000#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R4663), I initially used the `Txid` overload, but that fails the `p2p_orphan_handling.py::test_orphan_of_orphan` test. Switching to `Wtxid` makes it pass, but I don't understand why. Since with this pull, `TxOrphanage` indexes on `Wtxid`, calling `HaveTx` on a blind conversion of `parent_txid` into a `Wtxid` seems like it shouldn't find anything in the orphanage, unless we're dealing with non-segwit transaction?

Finally, for the [4) fourth here](https://github.com/glozow/bitcoin/compare/2024-04-orphan-use-wtxid...stickies-v:bitcoin:pr/30000#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R6319-R6341), I think my refactoring is a no-op, we just blindly call the different function signatures based on `GenTxid::IsWtxid()`, no real difference with the current implementation except perhaps for making things more explicit.

stickies-v replied:
To be clear: I'm not saying this PR needs to overhaul `AlreadyHaveTx` like I did, and I think having `TxOrphanage` index on `Wtxid` instead of `Txid` is better. I just think that if we're going to be calling `TxOrphanage` methods with (what are in fact) txids, we should probably have `TxOrphanage` _also_ keep a `Txid` index and not to blind conversion such as is happening here? I think that would address my concerns and make things easier to review?

instagibbs replied:
fwiw, I think the in-orphanage txid==wtixd case seems pretty rare all things considered. I had 4 over a 24 hour period, vs about 400 times where we fetched a segwit parent via txid though that particular txid was in the orphanage already.

stickies-v replied:
> iow I'm not sure extra complexity is worth it,

Oh, fair enough. I don't have a good enough view of the attack vectors here. In my review, I was operating on the assumption that we wouldn't want to miss anything we're currently doing. If it's okay to relax that, that could make sense - but it should probably be made explicit then?

instagibbs replied:
IIUC If the check is removed, you'll waste bandwidth when you get orphan chains that are non-segwit, but you won't keep or relay them, so it shouldn't churn the orphanage.

Would it help if the comment was touched up to be explicit about what it's saving us from doing?

itornaza replied:
imo `wtxid` and `txid` being double SHA256 outputs and `wtxid` being essentially a superset of `txid`, makes this blind conversion legitimate.

glozow replied:
Split up the commit, hope it helps (https://github.com/bitcoin/bitcoin/pull/30000#pullrequestreview-2049866191)

stickies-v reviewed (commented):
Concept ACK, but I'm [concerned](https://github.com/bitcoin/bitcoin/pull/30000/files#r1586713579) about the behaviour change in `AlreadyHaveTx`, which seems quite non-trivial to review so I'll need to look into further.

itornaza commented:
concept ACK

After being caught out of guard due to the short notice on this PR at the Bitcoin Review Club, I took time to review the code changes and read the comments above. 

To my understanding, we need to check the `wtxid` for screening transactions for inclusion in the orphanage in order to catch witness malleations as well. This was not previously possible because we where only checking against `txid` that does not include the `[witness]` section of a transaction as defined in [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#transaction-id). 

This change needs to be implemented only for the orphanage and not the rest of the structures that `AlreadyHaveTx()` checks against, since `txid` might be relevant there. The new behavior is implicitly documented in the `bool TxOrphanage::HaveTx(const Wtxid& wtxid) const` function definition which now takes `wtxid` as its single argument.

itornaza commented on src/net_processing.cpp:
```diff
@@ -2295,7 +2295,9 @@ bool PeerManagerImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconside
     const uint256& hash = gtxid.GetHash();
 
-    if (m_orphanage.HaveTx(gtxid)) return true;
+    // Orphanage is checked by wtxid. However, even if this is a txid, look up the same hash in
```
imo `wtxid` and `txid` being double SHA256 outputs and wtxid being essentially a superset of txid, makes this blind conversion legitimate.

itornaza commented on src/net_processing.cpp:
```diff
@@ -2295,7 +2295,9 @@ bool PeerManagerImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconside
-    if (m_orphanage.HaveTx(gtxid)) return true;
+    // Orphanage is checked by wtxid. However, even if this is a txid, look up the same hash in
+    // case this is a non-segwit transaction in the orphanage.
+    if (m_orphanage.HaveTx(Wtxid::FromUint256(gtxid.GetHash()))) return true;
```
imo wtxid and txid being double SHA256 outputs and wtxid being essentially a superset of txid, makes this blind conversion legitimate.

sr-gi reviewed (commented):
Concept ACK

glozow committed:
[p2p] don't query orphanage by txid

glozow committed:
[refactor] TxOrphanage::HaveTx only by wtxid

glozow commented on src/txorphanage.cpp:
```diff
@@ -23,7 +23,7 @@ bool TxOrphanage::AddTx(const CTransactionRef& tx, NodeId peer)
 
     const Txid& hash = tx->GetHash();
```
Yes, I would definitely say to log both txid and wtxid when we have it. A lot of things use txid only, e.g. if you want to trace a tx through logs or look it up in mempool. Would be a pain to not have txid imo.

glozow commented on src/txorphanage.cpp:
```diff
@@ -54,18 +52,19 @@ bool TxOrphanage::AddTx(const CTransactionRef& tx, NodeId peer)
+    std::map<Wtxid, OrphanTx>::iterator it = m_orphans.find(wtxid);
     if (it == m_orphans.end())
         return 0;
+    const auto& txid = it->second.tx->GetHash();
```
Same as above, imo definitely want to log txid if we have it.

glozow commented on src/txorphanage.cpp:
```diff
@@ -159,7 +156,7 @@ void TxOrphanage::AddChildrenToWorkSet(const CTransaction& tx)
+                std::set<Wtxid>& orphan_work_set = m_peer_work_set.try_emplace(elem->second.fromPeer).first->second;
                 // Add this tx to the work set
                 orphan_work_set.insert(elem->first);
                 LogPrint(BCLog::TXPACKAGES, "added %s (wtxid=%s) to peer %d workset\n",
```
Yes

glozow committed:
[refactor] TxOrphanage::EraseTx by wtxid

No behavior change right now, as transactions in the orphanage are
unique by txid. This makes the next commit easier to review.

glozow commented on src/test/fuzz/txorphan.cpp:
```diff
@@ -112,21 +112,21 @@ FUZZ_TARGET(txorphan, .init = initialize_orphanage)
                         if (ref) {
-                            bool have_tx = orphanage.HaveTx(GenTxid::Txid(ref->GetHash())) || orphanage.HaveTx(GenTxid::Wtxid(ref->GetWitnessHash()));
+                            bool have_tx = orphanage.HaveTx(ref->GetWitnessHash());
                             Assert(have_tx);
```
done

glozow commented on src/txorphanage.cpp:
```diff
@@ -226,7 +219,7 @@ void TxOrphanage::EraseForBlock(const CBlock& block)
                 const CTransaction& orphanTx = *(*mi)->second.tx;
-                const auto& orphanHash = orphanTx.GetHash();
+                const auto& orphanHash = orphanTx.GetWitnessHash();
                 vOrphanErase.push_back(orphanHash);
```
inlined. didn't do the rename as I don't think it's incorrect and want to minimize diff

glozow commented on src/txorphanage.h:
```diff
@@ -24,7 +24,7 @@ class TxOrphanage {
     bool AddTx(const CTransactionRef& tx, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
     /** Check if we already have an orphan transaction (by txid or wtxid) */
```
fixed thanks

glozow committed:
[unit test] TxOrphanage handling of same-txid-different-witness txns

glozow reviewed (commented):
Ok I have split the first commit into:
- net processing change to stop querying by txid
- 2 refactors changing the TxOrphanage methods to be by Wtxid only
- TxOrphanage change to index by wtxid and thus allow entries with the same txid

Hopefully in review, we can convince ourselves that the first commit is the right thing to do, separately from the other changes. The tests may help with that - also added one for invs that helps illustrate what the cases are, and added unit tests.

I thought a bit about whether we should check for same-txid-different-witness but from the same peer and replace that in orphanage. Given that a Bitcoin Core node doesn't ever do witness replacement, this seems like unnecessary complexity that won't ever get used between normally-operating nodes. So I did not add this.

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+        # 4. The parent is requested. Honest peer sends it.
+        honest_peer.wait_for_getdata([parent_txid_int])
+        honest_peer.send_message(msg_tx(tx_parent["tx"]))
+        honest_peer.sync_with_ping()
```
might want to note this line is necessary for the reconsideration step to always happen

glozow replied:
added comment

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+        # 1. Fake orphan is received first. It is missing an input.
+        bad_peer.send_and_ping(msg_tx(tx_orphan_bad_wit))
+
+        # 2. Node requests the missing parent by txid.
```
```Suggestion
        # 2. Node requests tx_grandparent by txid.
```

glozow replied:
done

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+        bad_peer.wait_for_getdata([grandparent_txid_int])
+
+        # 3. Honest peer relays the granchild, which is missing a parent. The parent by txid already
+        # exists in orphanage, but should be re-requested.
```
```Suggestion
        # exists in orphanage, but should be re-requested due to having witness data.
```

glozow replied:
elaborated

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+        with node.assert_debug_log(["stored orphan tx"]):
+            honest_peer.send_and_ping(msg_tx(tx_middle["tx"]))
+
+        # 5. The parent is requested. Honest peer sends it.
```
```Suggestion
        assert_equal(node.getrawmempool(), [])

        # 5. The parent is requested. Honest peer sends it.
```

glozow replied:
added

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+
+        # 5. The parent is requested. Honest peer sends it.
+        honest_peer.send_message(msg_tx(tx_grandparent["tx"]))
+        honest_peer.sync_with_ping()
```
is `send_message` + `sync_with_ping` always enough to resolve two generations of orphans to ensure we hit the mempool?

glozow replied:
I would say so. We do all `ProcessOrphanTx` before processing the ping; the child will be added to workset when the parent is accepted through `ProcessOrphanTx`

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+        node.bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        bad_peer.wait_for_getdata([parent_txid_int])
+
+        # 3. Honest peer announces the real parent, by txid (this isn't common but the node should
```
```Suggestion
        # 3. Honest peer announces the real child by txid (this isn't common but the node should
```

glozow replied:
gah thanks

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+
+    @cleanup
+    def test_orphan_txid_inv(self):
+        self.log.info("Check what happens when node receives announcement with same txid as tx in orphanage")
```
micronit to sound more prescriptive
```Suggestion
        self.log.info("Check node handles receiving announcement with same txid as tx in orphanage")
```

instagibbs replied:
note for readers: If the child tx ended up being the same wtxid, and in that case the node would simply not add to the orphanage or request anything further.

glozow replied:
changed to be more perscriptive

instagibbs reviewed (commented):
reviewed through 65f69bf533f66ecfb2695b03204bd437f1fe692d

> Given that a Bitcoin Core node doesn't ever do witness replacement, this seems like unnecessary complexity that won't ever get used between normally-operating nodes. So I did not add this.

Makes sense, seems unlikely to ever be used in a 20 minute window lifetime of an orphan given that we don't do these kinds of replacements.

also opened https://github.com/bitcoin/bitcoin/pull/30082 seeing we're touching `EraseForPeer` et al a bit in not quite "trivial" ways

theStack reviewed (commented):
Concept ACK

Code changes look good at first glance, still thinking about the implications of 9d2e9054134cedbf688368445eb6d2de14e2a8bd (w.r.t. discussion in https://github.com/bitcoin/bitcoin/pull/30000#discussion_r1586713579). Verified that both the unit tests and each of the newly introduced functional sub-tests fail without the main commit 4206ba3a36b329d1cc41df93cf59b685912d6d94, as expected.

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -57,10 +61,20 @@ def wrapper(self):
+        self._wtxidrelay = wtxidrelay
+
+    def on_version(self, message):
+        # Avoid sending verack in response to version.
```
Oops, this is a bad copypaste. I forgot to delete this commen I copied over.

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -57,10 +61,20 @@ def wrapper(self):
+
+    def on_version(self, message):
+        # Avoid sending verack in response to version.
+        # When calling add_p2p_connection, wait_for_verack=False must be set (see
```
Same - forgot to delete, sorry

glozow replied:
fixed

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+        node_mempool = node.getrawmempool()
+        assert tx_parent["txid"] in node_mempool
+        assert tx_child["txid"] in node_mempool
+        assert node.getmempoolentry(tx_child["txid"])["wtxid"] == tx_child["wtxid"]
```
done

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +428,152 @@ def test_orphan_inherit_rejection(self):
+        node_mempool = node.getrawmempool()
+        assert tx_parent["txid"] in node_mempool
+        assert tx_child["txid"] in node_mempool
+        assert node.getmempoolentry(tx_child["txid"])["wtxid"] == tx_child["wtxid"]
```
done

glozow committed:
[log] updates in TxOrphanage

- Add elapsed time in "remove orphan" log
- Add size in "stored orphan" log
- grammar edit

glozow commented on src/txorphanage.cpp:
```diff
@@ -101,13 +98,13 @@ void TxOrphanage::EraseForPeer(NodeId peer)
+            nErased += EraseTxNoLock(maybeErase->second.tx->GetWitnessHash());
         }
     }
     if (nErased > 0) LogPrint(BCLog::TXPACKAGES, "Erased %d orphan tx from peer=%d\n", nErased, peer);
```
Ok I've added a commit to fix this and another log to say "transaction(s)". I took the opportunity to add a few nice things to the logs as well.

glozow commented:
Thanks @instagibbs @theStack @sr-gi! Addressed the comments and added a few minor log improvements.

glozow commented on src/txorphanage.cpp:
```diff
@@ -101,13 +98,13 @@ void TxOrphanage::EraseForPeer(NodeId peer)
         if (maybeErase->second.fromPeer == peer)
         {
-            nErased += EraseTxNoLock(maybeErase->second.tx->GetHash());
+            nErased += EraseTxNoLock(maybeErase->second.tx->GetWitnessHash());
```
this was (mostly) taken

glozow commented:
Ah, p2p_invalid_tx.py failed because `assert_debug_log` didn't match anymore (those should be unit tests!)

instagibbs commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +427,154 @@ def test_orphan_inherit_rejection(self):
+
+        bad_peer = node.add_p2p_connection(PeerTxRelayer())
+        # Must not send wtxidrelay because otherwise the inv(TX) will be ignored later
+        honest_peer = node.add_p2p_connection(PeerTxRelayer(wtxidrelay=False))
```
I don't think either of these need to be `PeerTxRelayer`, just use `P2PInterface` and pass in `wtxidrelay` to remove the custom `on_version` handshake?

glozow replied:
Ah I missed that `P2PInterface` already had that! Thanks, deleted the unnecessary stuff.

instagibbs reviewed (commented):
reviewed through 16483fee7c6d93722bfb79fce9efbe841ec13d6a

`git range-diff master 65f69bf 16483fee7c6d93722bfb79fce9efbe841ec13d6a`

AngusP commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +427,154 @@ def test_orphan_inherit_rejection(self):
+        node.bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        bad_peer.wait_for_getdata([grandparent_txid_int])
+
+        # 3. Honest peer relays the granchild, which is missing a parent. The parent by txid already
```
```suggestion
        # 3. Honest peer relays the grandchild, which is missing a parent. The parent by txid already
```

AngusP reviewed (approved):
tACK 16483fee7c6d93722bfb79fce9efbe841ec13d6a with some nits

mzumsande commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +427,154 @@ def test_orphan_inherit_rejection(self):
+
+        # 4. The parent is requested. Honest peer sends it.
+        honest_peer.wait_for_getdata([parent_txid_int])
+        honest_peer.send_message(msg_tx(tx_parent["tx"]))
```
nit: could use `send_and_ping` here too (as in the other subtests).

mzumsande reviewed (commented):
Tested ACK 16483fee7c6d93722bfb79fce9efbe841ec13d6a

I was at first skeptical about the first commit for similar reasons as stickies-v above, but after thinking about it more and reading the discussion it makes sense to me to cast a `Txid` to a `Wtxid` here.
I also played with the functional tests a bit (e.g. giving the parent in `test_orphan_txid_inv` a low fee and testing that everything still works if 1p1c package validation is used).

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -398,6 +427,154 @@ def test_orphan_inherit_rejection(self):
+            honest_peer.send_and_ping(msg_tx(tx_middle["tx"]))
+        assert_equal(len(node.getrawmempool()), 0)
+
+        # 5. Hondest peer sends tx_grandparent
```
haha fixed

glozow commented on src/txorphanage.cpp:
```diff
@@ -47,7 +47,7 @@ bool TxOrphanage::AddTx(const CTransactionRef& tx, NodeId peer)
     }
 
-    LogPrint(BCLog::TXPACKAGES, "stored orphan tx %s (wtxid=%s) (mapsz %u outsz %u)\n", hash.ToString(), wtxid.ToString(),
+    LogPrint(BCLog::TXPACKAGES, "stored orphan tx %s (wtxid=%s), size: %u (mapsz %u outsz %u)\n", hash.ToString(), wtxid.ToString(), sz,
```
changed to "weight" !

glozow reviewed (commented):
Thanks @mzumsande @AngusP @instagibbs! Addressed comments. I've also expanded the "cast from txid to wtxid" part to be slightly more verbose but hopefully easier to understand.

instagibbs reviewed (approved):
ACK 0fb17bf61a40b73a2b81a18e70b3de180c917f22

reviewed via `git range-diff master 16483fee7c6d93722bfb79fce9efbe841ec13d6a 0fb17bf61a40b73a2b81a18e70b3de180c917f22`

glozow commented on src/net_processing.cpp:
```diff
@@ -2295,7 +2295,23 @@ bool PeerManagerImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconside
     const uint256& hash = gtxid.GetHash();
 
-    if (m_orphanage.HaveTx(gtxid)) return true;
+    if (gtxid.IsWtxid()) {
+        // Normal query by wtxid.
+        if (m_orphanage.HaveTx(Wtxid::FromUint256(hash))) return true;
+    } else {
+        // Never query by txid: it is possible that the transaction in the orphanage has the same
+        // txid but a different witness, which would give us a false positive result. If we decided
+        // not to request the transaction based on this result, an attacker could prevent us from
+        // downloading a transaction by intentionally creating a malleated version of it.  While
+        // only one (or none!) of these transactions can ultimately be confirmed, we have no way of
+        // discerning which one that is, so the orphanage can store multiple transactions with the
+        // same txid.
+        //
+        // While we won't query by txid, we can try to "guess" what the wtxid is based on the txid.
+        // A non-segwit transaction's txid == wtxid. Query this txid "casted" to a wtxid. This will
+        // help us find non-segwit transactions, saving bandwidth, and should have no false positives.
+        if (m_orphanage.HaveTx(Wtxid::FromUint256(hash))) return true;
+    }
```
(since @sr-gi mentioned this offline) Yes, the last push turned this 1 line into 2 identical lines. The purpose is to illustrate to current reviewers and future readers more clearly... if wtxid, we do this. if txid, we also do this, but notice that we are actually doing a "cast" and that's ok because we are *guessing* what the wtxid is.

Also, perhaps these will diverge again in the future if `GenTxid` becomes a `std::variant<Txid, Wtxid>`.

AngusP commented:
ACK 0fb17bf61a40b73a2b81a18e70b3de180c917f22

theStack reviewed (approved):
ACK 0fb17bf61a40b73a2b81a18e70b3de180c917f22

I like the latest, more verbose version of the first commit (7e475b9648bbee04f5825b922ba0399373eaa5a9), as it was significantly easier to grasp what's going on and convince myself that this is the behaviour we want. The only small drawback I see that it looks odd if both the if- and else-branch do exactly the same thing, and it could invite people to open deduplication refactor PRs (or leave them wondering if it's a bug). Maybe adding some "this is intended" comment would make sense for a follow-up.

sr-gi commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -394,10 +412,161 @@ def test_orphan_inherit_rejection(self):
+        honest_peer.send_and_ping(msg_tx(tx_parent["tx"]))
+
+        # 5. After parent is accepted, orphans should be reconsidered.
+        # The real child should be accepted and the fake one rejected.
```
In b16da7eda76944719713be68b61f03d4acdd3e16

The fake one is not really rejected, is it? It'll stay in the orphanage until a new block is mined (or until it times out)

sr-gi replied:
Oh my bad, I was thinking about children that may have multiple missing parents, not only the conflicting one.

sr-gi commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -394,10 +412,161 @@ def test_orphan_inherit_rejection(self):
+        honest_peer.wait_for_getdata([parent_txid_int])
+        honest_peer.send_and_ping(msg_tx(tx_parent["tx"]))
+
+        # 6. After parent is accepted, orphans should be reconsidered.
+        # The real child should be accepted and the fake one rejected. This may happen in either
+        # order since the message-processing is randomized. If tx_orphan_bad_wit is validated first,
+        # its consensus error leads to disconnection of bad_peer. If tx_child is validated first,
+        # tx_orphan_bad_wit is rejected for txn-same-nonwitness-data-in-mempool (no punishment).
```
In b16da7eda76944719713be68b61f03d4acdd3e16

Is this the case? Didn't we prioritize children from the same peer, so given the parent was given by the honest, and he also has presented a child, we will pick that one?

https://github.com/bitcoin/bitcoin/blob/dbb3113082a75035b14d20021036d2166171976e/src/net_processing.cpp#L3341-L3344

https://github.com/bitcoin/bitcoin/blob/dbb3113082a75035b14d20021036d2166171976e/src/net_processing.cpp#L3361-L3363

glozow replied:
You're referring to the 1p1c logic, which isn't executed here (notice that all of the feerates are "normal" feerates).
Here, the orphans are reconsidered with `ProcessOrphanTx`, which is called at the top of `ProcessMessages`.
https://github.com/bitcoin/bitcoin/blob/dbb3113082a75035b14d20021036d2166171976e/src/net_processing.cpp#L5355-L5369

`ProcessMessages` is called for each peer in a random order
https://github.com/bitcoin/bitcoin/blob/dbb3113082a75035b14d20021036d2166171976e/src/net.cpp#L2945-L2955

sr-gi replied:
Yep, my bad, nvm

AngusP commented:
> Maybe adding some "this is intended" comment would make sense for a follow-up.

Could make the 'cast' explicit in the second branch, though there's already the comment making it pretty clear this is intentional so may be unnecessary

```c++
const auto guessed_wtxid = Wtxid::FromUint256(hash);
if (m_orphanage.HaveTx(guessed_wtxid)) return true;
```

sr-gi reviewed (commented):
crACK [0fb17bf](https://github.com/bitcoin/bitcoin/pull/30000/commits/0fb17bf61a40b73a2b81a18e70b3de180c917f22)

Non-blocking comments

stickies-v reviewed (approved):
ACK 0fb17bf61a40b73a2b81a18e70b3de180c917f22

It took me a while to grok the implications of this PR, but the latest version of this PR makes it much more reviewable and I'm now comfortable that the changes are safe and desired. Also, nice work on the tests!

What I missed/misunderstood in my earlier review is that the reduced orphanage hit rate  is a goal, and not a bug/something to minimize. Related nit: perhaps the commit message of 7e475b9648bbee04f5825b922ba0399373eaa5a9 could be explicit about this reduction and why we want it?

I [remain uneasy](https://github.com/bitcoin/bitcoin/pull/30000#discussion_r1600159941) about `AlreadyHaveTx()` not really doing what the name indicates (because of the `GenTxid` ambiguity), but is not a huge concern and I think should not block this PR / can be addressed separately if it is desired.

itornaza commented:
trACK 0fb17bf61a40b73a2b81a18e70b3de180c917f22

ryanofsky commented:
I merged this as-is since it has so many acks, but it probably makes sense to follow up on the last few comments and suggestions.

glozow commented on src/txorphanage.cpp:
```diff
@@ -54,16 +54,19 @@ bool TxOrphanage::AddTx(const CTransactionRef& tx, NodeId peer)
+
+    std::map<Txid, OrphanTx>::iterator it = it_by_wtxid->second;
     if (it == m_orphans.end())
         return 0;
```
Good idea. Though it seems that this doesn't apply anymore since the commit is already merged, so resolving.

glozow commented on src/net_processing.cpp:
```diff
@@ -2295,7 +2295,20 @@ bool PeerManagerImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconside
+        // A non-segwit transaction's txid == wtxid. Query this txid "casted" to a wtxid. This will
+        // help us find non-segwit transactions, saving bandwidth, and should have no false positives.
+        if (m_orphanage.HaveTx(GenTxid::Wtxid(hash))) return true;
+    }
```
> I've summarized my understanding into the below

Looks correct to me

> I think the fact that we need such a verbose comment is a pretty strong sign AlreadyHaveTx should be refactored, e.g. using {Wt,T}xid types as I suggested https://github.com/bitcoin/bitcoin/pull/30000#discussion_r1587694422, but I appreciate that it's mostly orthogonal to this PR.

I think it's really a sign to ban txid-based transaction relay. Just kidding. I agree it's hard to reason about logic dealing with a hash that can be txid, wtxid, or bot. I think it's just one of the common things that trips people in tx relay, just like everybody's mempool fuzzer gets snagged on CPFP carve out, and that `for` loop in init.cpp that claims reviewer victims every time it's touched.

> Renaming AlreadyHaveTx to e.g. ShouldRequestTx or something would probably already clarify quite a bit, too (although also not blocking for this PR either).

Yeah, `AlreadyHaveTx` should probably have a name that reflects its purpose of "should I bother {requesting, downloading, validating} this tx".

Do note that #30110 makes `AlreadyHaveTx` a function internal to the `TxDownloadImpl`, i.e. no longer something that peerman knows about. Funnily, `TxDownloadImpl::AlreadyHaveTx` might be a fine name when it's in that context. I get that it's confusing, but I'd ask that we defer renaming until after that PR since it would conflict + maybe be less problematic afterwards.
